src/plugins/planout.ts

```diff
@@ -0,0 +1,364 @@
+    return result;
+  }
+
+  resetNamespaces() {
+    this.namespaces = {};
+  }
+
+  setOptions(_options: PlanoutPluginOptions) {
+    options = _options;
+  }
+
+  setInput(key: string, value: ParamValue) {
+    if (this.inputs[key] !== value) {
+      this.inputs[key] = value;
+      ExperimentSetup.registerExperimentInput(key, value);
+      Object.keys(this.namespaces).forEach((namespaceName: string) => {
+        const ns = this.namespaces[namespaceName];
+        this.namespaces[namespaceName] = ns.clone();
+      });
+    }
+  }
+
+  addNamespace(namespace: string, numSegments: number = 1000, primaryUnit: string = 'uid', compiled?: PlanoutCode) {
+    if (!this.namespaces[namespace]) {
+      const ns = new BaseNamespace(namespace);
+      this.namespaces[namespace] = ns;
+      ns.setName(namespace);
+      ns.numSegments = numSegments;
+      ns.setPrimaryUnit(primaryUnit);
+      ns.setDefaultExperimentConfig({
+        namespace_name: namespace,
+        name: '_default',
+        compiled: compiled || { op: 'seq', seq: [] },
+        num_segments: 0,
+      });
+    }
+    return this.namespaces[namespace];
+  }
+
+  add(experimentConfig: PlanoutExperimentConfig) {
+    const namespace = experimentConfig.namespace_name;
+    if (!this.namespaces[namespace]) {
+      this.addNamespace(namespace);
+    }
+    this.namespaces[namespace].addExperimentConfig(experimentConfig);
+  }
+
+  remove(namespace: string, name: string) {
+    if (this.namespaces[namespace]) {
+      this.namespaces[namespace].removeExperiment(name);
+    }
+  }
+
+  useConfig(config: PlanoutConfig) {
+    this.resetNamespaces();
+    config.namespaces.forEach((nsConfig) => {
+      this.addNamespace(
+        nsConfig.name,
+        nsConfig.num_segments,
+        nsConfig.primary_unit,
+        nsConfig.compiled);
+    });
+
+    config.experiments.forEach((nsOp) => {
+      const experiment = nsOp.experiment;
+      if (nsOp.op === 'add') {
+        this.add(experiment);
+      } else if (nsOp.op === 'remove') {
+        this.remove(experiment.namespace_name, experiment.name);
+      } else {
+        console.error(`Unknown operation ${nsOp.op}.`);
+      }
+    });
+
+    if (options.localOverrides) {
+      Object.keys(options.localOverrides).forEach((namespace: string) => {
+        const nsOverrides = options.localOverrides![namespace];
+        const ns = this.namespaces[namespace] || this.addNamespace(namespace);
+        nsOverrides.forEach((override) => {
+          ns.setOverride('default', override.variable, override.value);
+        });
+      });
+    }
+  }
+
+  async fetchConfig() {
+    try {
+      const response = await this.axios.get<PlanoutConfig>(`/cms/planout/config/${options.appName}`);
+      const config = await options.getConfig();
+      if (!isEqual(response.data, config)) {
+        console.info('Planout: Experiment config updated. Re-initializing.');
+        options.setConfig(response.data);
+        this.useConfig(response.data);
```

Comment:
A bit confused about this line as well. Wouldn't we always want to `useConfig` when initializing the plugin?

Reply:
Still confused... i can't find where this is called now

Reply:
Refactored (well you did)